### PR TITLE
model: add UNITE (friedrichor/Unite-Base-Qwen2-VL-2B)

### LINKED ISSUE #5185: Add model: UNITE (friedrichor/Unite-Base-Qwen2-VL-2B)
Claimed in #4842 (Video row).

**UNITE** is a universal multimodal embedder over text, image, and video, plus combined queries.

- Paper: Modality Curation: Building Universal Embeddings for Advanced Multimodal Information Retrieval, arXiv:2505.19650 (Kong et al., Northeastern University and Kuaishou)
- Code: https://github.com/friedrichor/UNITE
- Weights: https://huggingface.co/collections/friedrichor/unite-682da30c4540abccd3da3a6b

**Details**

Qwen2-VL backbone fine-tuned for retrieval with Modal-Aware Masked Contrastive Learning. Embeddings come from E5-V style prompting ("Summary above sentence in one word:"), last-token pooling, L2 normalized. `friedrichor/Unite-Base-Qwen2-VL-2B` outputs 1536-d. There are both Base and Instruct families in the collection; Base is recommended for fine-grained retrieval and Instruct for instruction-based retrieval.

Covers t2i/i2t, t2v/v2t, and composed image+text queries, so it can run across the image and video parts of the matrix rather than one direction.

**Implementation note**

The custom class lives in `modeling_unite.py` and only subclasses `Qwen2VLForConditionalGeneration`. There is no `auto_map` on the HF repos, so `trust_remote_code` is not an option and the class needs to be vendored into the wrapper. Upstream requirements pin `transformers==4.47.1`, but I have it working on 4.52.4 with a two-line change (`embed_tokens` and the decoder call moved under `model.language_model` in the Qwen2VL refactor), so no version pin should be needed. Weight remapping is handled by the existing `_checkpoint_conversion_mapping`.

Planning to reproduce results on 1-2 benchmarks from the paper (CaReBench or WebVid-CoVR) as the reference run.

### PR DESCRIPTION
**Pull request by hubielu**
Closes #5185. Claimed in #4842 (Video row).

Adds UNITE (arXiv:2505.19650), a universal embedder for text, image, video, and fused queries. Qwen2-VL-2B backbone, E5-V style prompting, last-token pooling, L2 normalization, 1536-d embeddings.

## Reference run

CaReBench, using 32-frame video sampling:

| Task | nDCG@10 | R@1 | R@5 | R@10 |
|---|---:|---:|---:|---:|
| CaReBenchT2VRetrieval | 89.03 | 79.00 | 95.80 | 97.80 |
| CaReBenchV2TRetrieval | 90.13 | 80.30 | 96.50 | 98.60 |

`Unite-Base-Qwen2-VL-2B` @ `850056e7`, bf16, 1x A6000, 32 sampled frames, video capped at `360*420`.

MTEB uses the CaRe-General queries. The paper reports:

- T→V: 78.1 / 95.5 / 97.8
- V→T: 80.8 / 96.4 / 98.7

for R@1/R@5/R@10, so the reproduction is within 0.9 R@1 points and essentially matches R@5/R@10. The earlier gap came from fps-based sampling.

Separately reproduced upstream demo similarities closely: text-image 0.7539 (0.7500 reported), text-video 0.5625 (0.5664 reported).

## Note on loading

UNITE ships `UniteQwen2VL`, a `Qwen2VLForConditionalGeneration` subclass, without an `auto_map`.

On transformers 5.x, the base class loaded the checkpoint correctly, but the UNITE subclass did not receive the same key remapping and loaded only 391/729 weights, producing invalid embeddings.

The final implementation restores UNITE's subclass/forward behavior and uses MTEB's existing `transformers-v4` requirement group.

## Video / batching

Video resizing is applied through the processor with `max_pixels=360*420`. Default sampling is 32 frames, while `fps`, `max_frames`, and `num_frames` remain configurable.

Batch encoding is supported. On 16 CaReBench videos, bs=1 was fastest (3.59 s/item vs. 3.92/4.21/5.16 for bs=2/4/8), so batch size remains configurable.

## Follow-up

Instruct checkpoints can be added separately as additional `ModelMeta` entries.

### CODE REVIEW

File: mteb/models/model_implementations/unite_models.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    scale = (max_pixels / (h * w)) ** 0.5
+    nh = max(28, int(h * scale) // 28 * 28)
+    nw = max(28, int(w * scale) // 28 * 28)
+    return (
+        torch.nn.functional.interpolate(
+            video.float(), size=(nh, nw), mode="bilinear", align_corners=False
+        )
+        .clamp(0, 255)
+        .to(torch.uint8)
+    )
```
Comment (Samoed) [inline mteb/models/model_implementations/unite_models.py]: I think this will be done in processor https://github.com/huggingface/transformers/blob/60eefbc8141a028609875951f5170086631d9605/src/transformers/models/qwen2_vl/video_processing_qwen2_vl.py#L295-L307

File: mteb/models/model_implementations/unite_models.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    # language model silently loads with random weights. We keep the base class and
+    # reproduce UNITE's pooling in _unite_embed instead.
+    try:
+        return Qwen2VLForConditionalGeneration.from_pretrained(
+            model_name, revision=revision, dtype=torch.bfloat16
+        )
+    except TypeError:
+        return Qwen2VLForConditionalGeneration.from_pretrained(
+            model_name, revision=revision, torch_dtype=torch.bfloat16
+        )
```
Comment (Samoed) [inline mteb/models/model_implementations/unite_models.py]: How this solves this problem? Maybe would be easier to pin transformers version?

File: mteb/models/model_implementations/unite_models.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        video_max_pixels: int = 360 * 420,
+        **kwargs: Any,
+    ) -> None:
+        from transformers import AutoProcessor
+
+        self.device = device or ("cuda" if torch.cuda.is_available() else "cpu")
+        self.fps = fps
+        self.max_frames = max_frames
+        self.target_sampling_rate = target_sampling_rate
+        self.video_max_pixels = video_max_pixels
```
Comment (Samoed) [inline mteb/models/model_implementations/unite_models.py:141]: Shoudn't this be in processor?

File: mteb/models/model_implementations/unite_models.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+            texts = batch.get("text")
+            images = batch.get("image")
+            videos = batch.get("video")
+            n = len(texts or images or videos)
+            for i in range(n):
+                emb = self._embed_one(
+                    text=texts[i] if texts is not None else None,
+                    image=images[i] if images is not None else None,
+                    video=videos[i] if videos is not None else None,
+                )
```
Comment (Samoed) [inline mteb/models/model_implementations/unite_models.py]: Why your're not batch this?

File: mteb/models/model_implementations/unite_models.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    reference="https://huggingface.co/friedrichor/Unite-Base-Qwen2-VL-2B",
+    similarity_fn_name=ScoringFunction.COSINE,
+    framework=["PyTorch", "safetensors", "Transformers"],
+    use_instructions=False,
+    public_training_code="https://github.com/friedrichor/UNITE",
+    public_training_data="https://huggingface.co/datasets/friedrichor/Unite-Base-Retrieval-Train",
+    training_datasets=unite_training_datasets,
+    adapted_from="Qwen/Qwen2-VL-2B-Instruct",
+    citation=UNITE_CITATION,
+    contacts=["hubielu"],
```
Comment (Samoed) [inline mteb/models/model_implementations/unite_models.py]: There should be model's authors
```suggestion
```

File: mteb/models/model_implementations/unite_models.py (hunk 1 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: You're right, this belongs in the processor. I'll drop `_fit` and configure the video processor's size instead

File: mteb/models/model_implementations/unite_models.py (hunk 2 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: The remapping is applied to the known class, so a subclass doesn't get it. On 5.15.0: base loads 729/729, subclass 391/729 silently, and upstream's demo similarity drops 0.7500 → 0.056.

Pinning is awkward since UNITE wants 4.47.1 and mteb resolves 5.15.0. Base class needs no pin, verified on 4.52.4 and 5.15.0.

The try/except is only for the torch_dtype → dtype rename; happy to drop it if the floor already covers it.

File: mteb/models/model_implementations/unite_models.py (hunk 3 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py:141]: video_max_pixels yes. fps and max_frames go to VideoCollator, not the processor, so those stay unless you'd rather they be loader_kwargs.

File: mteb/models/model_implementations/unite_models.py (hunk 4 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: No good reason. Will batch and re-run CaReBench to confirm scores hold.

File: mteb/models/model_implementations/unite_models.py (hunk 2 above)
Comment (Samoed) [reply mteb/models/model_implementations/unite_models.py]: > Pinning is awkward since UNITE wants 4.47.1 and mteb resolves 5.15.0

We don't pin high transformers version. A lot of models pin lower version of transformers, e.g. https://github.com/embeddings-benchmark/mteb/blob/687d5ecad9e2f57d94beeaefd94077c9140a7125/pyproject.toml#L147-L148

File: mteb/models/model_implementations/unite_models.py (hunk 4 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: Batched! Measured on 16 CaReBench videos (A6000, fps=1, max_frames=32): bs=1 3.59s/item, bs=2 3.92, bs=4 4.21, bs=8 5.16 — clips run 9-32 frames, so padding to batch max costs more than it saves. Left to the CLI.

Re-ran at bs=1, with the video cap now in the processor:

| Task | nDCG@10 | R@1 |
|---|---|---|
| CaReBenchT2VRetrieval | 88.36 | 77.70 |
| CaReBenchV2TRetrieval | 89.62 | 79.40 |

Was 87.41/76.00 and 90.15/80.40. The ~1pt drift is the processor's smart-resize landing on a different grid than the manual resize, same 360*420 budget.

File: mteb/models/model_implementations/unite_models.py (hunk 2 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: That makes sense. Though I don't think a pin is needed here either way since the base class loads fine on both 4.x and 5.x, so pinning <5 would only be to keep the subclass. Happy to pin instead if you'd prefer that shape.

File: mteb/models/model_implementations/unite_models.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    # language model silently loads with random weights. We keep the base class and
+    # reproduce UNITE's pooling in _unite_embed instead.
+    try:
+        return Qwen2VLForConditionalGeneration.from_pretrained(
+            model_name, revision=revision, dtype=torch.bfloat16
+        )
+    except TypeError:
+        return Qwen2VLForConditionalGeneration.from_pretrained(
+            model_name, revision=revision, torch_dtype=torch.bfloat16
+        )
```
Comment (Samoed) [inline mteb/models/model_implementations/unite_models.py]:
```suggestion
    return Qwen2VLForConditionalGeneration.from_pretrained(
        model_name, revision=revision, torch_dtype=torch.bfloat16
    )
```

File: mteb/models/model_implementations/unite_models.py (hunk 2 above)
Comment (Samoed) [reply mteb/models/model_implementations/unite_models.py]: But you said 
> On 5.15.0: base loads 729/729, subclass 391/729 silently, and upstream's demo similarity drops 0.7500 → 0.056.

So transformers can't load this model correctly on v5 version or I miss something?

File: mteb/models/model_implementations/unite_models.py (hunk 4 above)
Comment (Samoed) [reply mteb/models/model_implementations/unite_models.py]: Seems perforamcne is a bit lower that in paper https://arxiv.org/pdf/2505.19650. Can you investigate this a bit more why this happen? Would be great to reproduce scores

File: mteb/models/model_implementations/unite_models.py (hunk 2 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: Sorry, that was unclear. transformers loads it fine on v5, the base class gets 729/729. Only the subclass fails, because the key remapping is registered against the known class and doesn't reach a user subclass. So no pin is needed; the wrapper just uses the base class.

File: mteb/models/model_implementations/unite_models.py (hunk 4 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: The paper's Table 1 seems to report R@1 separately for CaRe-General / Spatial / Temporal, but mteb ships CaReBench as a single task, so I'm not sure which row I should be comparing against. Do you know which one it's meant to match?

My best guess at the gap is frame sampling. I'm using fps=1 to match UNITE's inference demo, but CaReBench's own paper says their experiments use 32 uniformly sampled frames, and since the clips are mostly 5-30s, fps=1 probably gives fewer than that. Could easily be something else though. I'll try num_frames=32 and post both sets of numbers.

Comment (Samoed) [reply mteb/models/model_implementations/unite_models.py]: > CaRe-General / Spatial / Temporal, but mteb ships CaReBench as a single task, so I'm not sure which row I should be comparing against

These are just different queries. We have general implementation

File: mteb/models/model_implementations/unite_models.py (hunk 2 above)
Comment (Samoed) [reply mteb/models/model_implementations/unite_models.py]: > Only the subclass fails, because the key remapping is registered against the known class and doesn't reach a user subclass

Which subclass fails? Shoudn't we pin then?

Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: The failing subclass is UNITE's UniteQwen2VL(Qwen2VLForConditionalGeneration), vendored from their inference/modeling_unite.py. Agreed on constraining the transformers version. I restored that subclass and added MTEB's existing transformers-v4 requirement group in 7b713204, so this model stays off transformers v5 instead of using the base-class workaround.

File: mteb/models/model_implementations/unite_models.py (hunk 3 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py:141]: video_max_pixels is now applied through the processor. fps and max_frames stay on VideoCollator since they control frame sampling rather than processor resizing.

File: mteb/models/model_implementations/unite_models.py (hunk 4 above)
Comment (hubielu) [reply mteb/models/model_implementations/unite_models.py]: Thanks, that clarifies it. I compared against the CaRe-General row.

I reran CaReBench with 32 uniformly sampled frames. For UNITEbase 2B I get:

T→V: R@1 79.0, R@5 95.8, R@10 97.8
V→T: R@1 80.3, R@5 96.5, R@10 98.6

The paper reports CaRe-General:

T→V: 78.1 / 95.5 / 97.8
V→T: 80.8 / 96.4 / 98.7

So the reproduction is within 0.9 R@1 points and essentially exact at R@5/R@10. The earlier discrepancy was sensitive to the video sampling setup.

File: mteb/models/model_implementations/unite_models.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+        with torch.inference_mode():
+            return self.model(**inputs)
+
+    def _embed_one(self, text=None, image=None, video=None) -> torch.Tensor:
+        """Single-item convenience wrapper, used by tests."""
+        return self._embed_batch(
+            texts=[text] if text is not None else None,
+            images=[image] if image is not None else None,
+            videos=[video] if video is not None else None,
+        )
```
Comment (Samoed) [inline mteb/models/model_implementations/unite_models.py]: Not used
```suggestion
```

**Review by Samoed — APPROVED** [review]
